### ISSUE TITLE
What was the fix for support of SELECT * FROM s3(...) across versions ?

### ISSUE THREAD

**Issue by Amogh-Bharadwaj** [ClickHouse/ClickHouse#63573]
### TL;DR
I am noticing that the query of table function:
```
SELECT * FROM s3(.., 'Avro') LIMIT 1;
```
raises a server exception in Clickhouse version 23.8.12.13, but it works in version 24.3.2. 

I would **like to understand the bug fix / improvement that was done** between these versions for this feature.

The error is:
```
Syntax error (columns declaration list): failed at position 5 (end of query): . 
Expected one of: NULL, NOT, DEFAULT, MATERIALIZED, EPHEMERAL, ALIAS, AUTO_INCREMENT, PRIMARY KEY, data type, nested table, identifier. 
(SYNTAX_ERROR)
```

### Context
I'm using docker images of clickhouse of these 2 versions, along with clickhouse client.
Clickhouse client version: 
```
ClickHouse client version 24.2.1.275 (official build).
```
The bucket I am querying from is a [https://min.io/](minio) bucket, hosted locally using a min.io docker container.

### Details of the query result in 23.8.12.13
```sh
c2a5da5eb070 :) 
SELECT *
FROM s3(
    'http://host.docker.internal:9001/mybucket/path_to_avro.zst',
    'access_key_id',
    'secret_access_key',
    '',
    'Avro'
) LIMIT 1;

Query id: 8f054aa7-de0c-42cb-b8f5-d7bcf14a7457

Elapsed: 0.006 sec. 

Received exception from server (version 23.8.12):
Code: 62. DB::Exception: Received from localhost:9005. DB::Exception: 
Syntax error (columns declaration list): failed at position 5 (end of query): . 
Expected one of: NULL, NOT, DEFAULT, MATERIALIZED, EPHEMERAL, ALIAS, AUTO_INCREMENT, PRIMARY KEY, data type, nested table, identifier. 
(SYNTAX_ERROR)
```

Here is the stack trace obtained from clickhouse.log:
```
2024.05.09 14:46:08.672135 [ 48 ] {3b466603-0fc1-4f3e-b59a-7ce981dd6de1} <Debug> executeQuery: (from 172.17.0.1:57720) SELECT * FROM s3(<redacted>,'Avro') LIMIT 1 (stage: Complete)
2024.05.09 14:46:08.672313 [ 48 ] {3b466603-0fc1-4f3e-b59a-7ce981dd6de1} <Trace> NamedCollectionsUtils: Loaded 0 collections from SQL
2024.05.09 14:46:08.672415 [ 48 ] {3b466603-0fc1-4f3e-b59a-7ce981dd6de1} <Trace> ContextAccess (default): Access granted: CREATE TEMPORARY TABLE, S3 ON *.*
2024.05.09 14:46:08.673860 [ 48 ] {3b466603-0fc1-4f3e-b59a-7ce981dd6de1} <Error> executeQuery: Code: 62. DB::Exception: Syntax error (columns declaration list): failed at position 5 (end of query): . Expected one of: NULL, NOT, DEFAULT, MATERIALIZED, EPHEMERAL, ALIAS, AUTO_INCREMENT, PRIMARY KEY, data type, nested table, identifier. (SYNTAX_ERROR) (version 23.8.12.13 (official build)) (from 172.17.0.1:57720) (in query: SELECT * FROM s3(<redacted>, '', 'Avro') LIMIT 1), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000bfc4ec4 in /usr/bin/clickhouse
1. DB::Exception::createDeprecated(String const&, int, bool) @ 0x0000000007969320 in /usr/bin/clickhouse
2. DB::parseQueryAndMovePosition(DB::IParser&, char const*&, char const*, String const&, bool, unsigned long, unsigned long) @ 0x00000000128f72f8 in /usr/bin/clickhouse
3. DB::parseColumnsListFromString(String const&, std::shared_ptr<DB::Context const> const&) @ 0x00000000114e720c in /usr/bin/clickhouse
4. DB::TableFunctionS3::executeImpl(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool) const @ 0x000000000fdb77e8 in /usr/bin/clickhouse
5. DB::ITableFunction::execute(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context const>, String const&, DB::ColumnsDescription, bool, bool) const @ 0x000000000fffaf8c in /usr/bin/clickhouse
6. DB::Context::executeTableFunction(std::shared_ptr<DB::IAST> const&, DB::ASTSelectQuery const*) @ 0x00000000109d0684 in /usr/bin/clickhouse
7. DB::JoinedTables::getLeftTableStorage() @ 0x0000000011210f5c in /usr/bin/clickhouse
8. DB::InterpreterSelectQuery::InterpreterSelectQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context> const&, std::optional<DB::Pipe>, std::shared_ptr<DB::IStorage> const&, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&, std::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::shared_ptr<DB::PreparedSets>) @ 0x0000000011140ae4 in /usr/bin/clickhouse
9. DB::InterpreterSelectWithUnionQuery::InterpreterSelectWithUnionQuery(std::shared_ptr<DB::IAST> const&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&, std::vector<String, std::allocator<String>> const&) @ 0x00000000111d5648 in /usr/bin/clickhouse
10. DB::InterpreterFactory::get(std::shared_ptr<DB::IAST>&, std::shared_ptr<DB::Context>, DB::SelectQueryOptions const&) @ 0x0000000011102aa4 in /usr/bin/clickhouse
11. DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum, DB::ReadBuffer*) @ 0x00000000114c5d24 in /usr/bin/clickhouse
12. DB::executeQuery(String const&, std::shared_ptr<DB::Context>, bool, DB::QueryProcessingStage::Enum) @ 0x00000000114c2c88 in /usr/bin/clickhouse
13. DB::TCPHandler::runImpl() @ 0x00000000121bce98 in /usr/bin/clickhouse
14. DB::TCPHandler::run() @ 0x00000000121ccfe8 in /usr/bin/clickhouse
15. Poco::Net::TCPServerConnection::start() @ 0x00000000146dac84 in /usr/bin/clickhouse
16. Poco::Net::TCPServerDispatcher::run() @ 0x00000000146dc180 in /usr/bin/clickhouse
17. Poco::PooledThread::run() @ 0x000000001484f2fc in /usr/bin/clickhouse
18. Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001484d4a4 in /usr/bin/clickhouse
19. start_thread @ 0x0000000000007624 in /usr/lib/aarch64-linux-gnu/libpthread-2.31.so
20. ? @ 0x00000000000d162c in /usr/lib/aarch64-linux-gnu/libc-2.31.so```
```

### Query result for v24.3.2
The query returns successfully

Query id: 8ed2c4ca-a47e-487e-aa70-d7c8c6d91f6f

┌────────id─┬──────c1─┬──────c2─┬─t─────────────────────────────────────┬─────────────────updated_at─┐
│ 160857778 │ 2660146 │ 3504610 │helloworld │ 2024-05-06 10:58:12.234476 │
└───────────┴─────────┴─────────┴───────────────────────────────────────┴────────────────────────────┘

1 row in set. Elapsed: 0.055 sec. 

8e11b3d5910b :) Bye.
```
